container/service.go

```diff
@@ -67,38 +67,44 @@ func (c *DockerContainer) StopService(namespace []string) error {
 	if status == STOPPED {
 		return nil
 	}
-
+	service, err := c.FindService(namespace)
+	if err != nil && !docker.IsErrNotFound(err) {
+		return err
+	}
+	stopGracePeriod := c.defaultStopGracePeriod
+	if service.Spec.TaskTemplate.ContainerSpec != nil && service.Spec.TaskTemplate.ContainerSpec.StopGracePeriod != nil {
+		stopGracePeriod = *service.Spec.TaskTemplate.ContainerSpec.StopGracePeriod
+	}
 	ctx, cancel := context.WithTimeout(context.Background(), c.callTimeout)
 	defer cancel()
 	if err := c.client.ServiceRemove(ctx, c.Namespace(namespace)); err != nil && !docker.IsErrNotFound(err) {
 		return err
 	}
-	if err := c.deletePendingContainer(namespace); err != nil {
+	if err := c.deletePendingContainer(namespace, time.Now().Add(stopGracePeriod)); err != nil {
 		return err
 	}
 	return c.waitForStatus(namespace, STOPPED)
 }
 
-func (c *DockerContainer) deletePendingContainer(namespace []string) error {
-	ctx, cancel := context.WithTimeout(context.Background(), c.callTimeout)
-	defer cancel()
+func (c *DockerContainer) deletePendingContainer(namespace []string, maxGraceTime time.Time) error {
```

Comment:
We can simplify time logic by using [<-time.After(stopGracePeriod)](https://golang.org/pkg/time/#After).

Reply:
I don't see how it will simplify.

The goal here is to "return" as soon as the container is not found.
The maxGraceTime is kind of a timeout.

I think using `<-time.After(stopGracePeriod)` will always to call `ContainerRemove` or to prevent this, another chan done need to be created to tell the go routine to stop. So actually it seems more complex.

Reply:
yes, i guess current implementation is fine because after we remove the service and check if the container exists right after, Docker still might be in the deleting process of container. So it's good that we don't block for the whole timeout and check the status of container again in short periods.